x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RemoteAccessAuthenticationService.java

```diff
@@ -0,0 +1,252 @@
+import java.util.Set;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.core.Strings.format;
+import static org.elasticsearch.xpack.core.security.authc.RemoteAccessAuthentication.REMOTE_ACCESS_AUTHENTICATION_HEADER_KEY;
+import static org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor.REMOTE_ACCESS_CLUSTER_CREDENTIAL_HEADER_KEY;
+
+public class RemoteAccessAuthenticationService {
+
+    public static final RoleDescriptor CROSS_CLUSTER_INTERNAL_ROLE = new RoleDescriptor(
+        "_cross_cluster_internal",
+        new String[] { ClusterStateAction.NAME },
+        null,
+        null,
+        null,
+        null,
+        null,
+        null,
+        null
+    );
+    private static final Logger logger = LogManager.getLogger(RemoteAccessAuthenticationService.class);
+    private final ClusterService clusterService;
+    private final ApiKeyService apiKeyService;
+    private final AuthenticationService authenticationService;
+
+    public RemoteAccessAuthenticationService(
+        ClusterService clusterService,
+        ApiKeyService apiKeyService,
+        AuthenticationService authenticationService
+    ) {
+        this.clusterService = clusterService;
+        this.apiKeyService = apiKeyService;
+        this.authenticationService = authenticationService;
+    }
+
+    public void authenticate(
+        final String action,
+        final TransportRequest request,
+        final boolean allowAnonymous,
+        final ActionListener<Authentication> listener
+    ) {
+        final Authenticator.Context authcContext = authenticationService.newContext(action, request, allowAnonymous);
+        final ThreadContext threadContext = authcContext.getThreadContext();
+
+        // TODO revisit this once Node's Version is refactored
+        if (getMinNodeVersion().before(Authentication.VERSION_REMOTE_ACCESS_REALM)) {
+            withRequestProcessingFailure(
+                authcContext,
+                new IllegalArgumentException(
+                    "all nodes must have version ["
+                        + Authentication.VERSION_REMOTE_ACCESS_REALM
+                        + "] or higher to support cross cluster requests with remote access"
+                ),
+                listener
+            );
+            return;
+        } else if (threadContext.getHeader(AuthenticationField.AUTHENTICATION_KEY) != null) {
+            withRequestProcessingFailure(
+                authcContext,
+                new IllegalArgumentException("authentication header is not allowed with remote access"),
+                listener
+            );
+            return;
+        } else if (threadContext.getHeader(AuthenticationServiceField.RUN_AS_USER_HEADER) != null) {
+            withRequestProcessingFailure(
+                authcContext,
+                new IllegalArgumentException("run-as header is not allowed with remote access"),
+                listener
+            );
+            return;
```

Comment:
I think we should have an allowlist instead of a denylist for the headers similar to the HTTP interface. At the HTTP interface, unrecognised headers are simply ignored (not copied to threadContext). The remote access port is conceptually like the HTTP interface. It should either reject or drop any unrecognised headers as well.

Reply:
I've added an allowlist but moved it to the transport filter -- it does not feel like a service-level concern but something that should be handled further up the stack. 

I'm sticking with rejecting for now, instead of copying headers. In the long run, copying makes sense but explicitly failing will help us discover headers (I'm thinking things like audit request ID) that we might be missing more quickly.

Reply:
Yeah have it in the filter is good. The HTTP headers are also handled closer to the network layer. 

> I'm sticking with rejecting for now, instead of copying headers. In the long run, copying makes sense but explicitly failing will help us discover headers (I'm thinking things like audit request ID) that we might be missing more quickly.

I wasn't suggesting copying. The headers are already in threadContext. So there is no need to copy anyway? By rejecting, do you mean denylist or reject if it is not in allowlist? 

Besides audit ID, there are [a few more headers](https://github.com/elastic/elasticsearch/blob/ea0d1a315f5324e024b39f8a5593170c0da88b32/server/src/main/java/org/elasticsearch/tasks/Task.java#L59-L64) we need to allow. 

Allowlist is my preference. But I do recognise the risk that if there is any new header added by other team without us being aware of it, it will break user's CCS setup and there is no easy way to fix other than upgrade. So maybe we should either (1) go with denylist or (2) apply the allowlist on the QC side (so it can drop new headers but at least won't break the CCS).

Reply:
Sorry, my comment was unclear: I meant rejecting requests as opposed to only copying over allowed headers before processing. So exactly as you suggest, an allowlist.

Reply:
I'll add the extra headers you mentioned to the allowlist, and add a Jira issue to revisit making this more robust to header changes. I'm leaning towards having the QC apply the allowlist.